### ISSUE TITLE
Automated check for detecting missing localizations

### ISSUE THREAD

**Issue by simonharrer** [JabRef/jabref#436]
At the moment, the python script `scripts/syncLang.py` is used to detect issues with the localization. We want to run this check as part of the CircleCI build, and eventually also part of the gradle build locally.

For the CircleCI build, it would be ok for starts to install python within the CircleCI container and then run the script. The script may need to be altered to return a status code != 0 when the localization files are in an inconsistent state. 

For the gradle build, we should use jython which can be loaded by gradle itself. This would allow to execute the script without having to install a separate runtime, reducing the burden of new developers to create valid pull requests.

**Comment by koppor** [comment]
I'm currently investigating whether [gettext-commons](https://code.google.com/p/gettext-commons/) would solve much of our issues.

**Comment by simonharrer** [comment]
Gradle language tasks implemented in #403 with an independent gradle script, which can be executed by the gradle wrapper. 

@matthiasgeiger would this fit the purpose? Maybe you can help with naming of the tasks? They have quite long names...

**Comment by matthiasgeiger** [comment]
e.g. https://github.com/JabRef/jabref/blob/master/src/main/java/net/sf/jabref/migrations/FileLinksUpgradeWarning.java#L65

**Comment by matthiasgeiger** [comment]
Okay... finally found some time to check it out and test it... Works fine with gradle...

Two aspects are crucial for  the localization: 
- There should be no missing keys - i.e., all used Strings should be added to `JabRef_en.properties`
- All keys listed in `JabRef_en.properties` should also be available for the other languages (necessary for the translators to know what to translate)

Obsolete keys and duplicates are not that crucial - but should be cleaned prior to releases.
So for us developers it would be nice if the creation of the keys is as easy as possible. 
Status quo: 
  1. Run `syncLang.py -s` to check whether some keys are missing (output is cluttered with lots of "obsolete keys")
  2. Check whether keys can be reused, typos in keys, ...? 
  3. If needed:  Run `syncLang.py -s -u` create really new keys for english
  4. If needed: Run `syncLang.py -t -u` create the newly created keys for the other languages

I think a good workflow would be:
  1. Add a Unit test which automatically performs step 1) Are any keys missing? If so, fail.
  2. Check why the test fails: Can keys be reused, typos...?
  3. If needed: Run "gradlew -b localization.gradle createMissingKeys" (which performs step 3. and 4. from above)

What do you think?

**Comment by koppor** [comment]
:+1:

**Comment by simonharrer** [comment]
Who does "Check whether keys can be reused, typos in keys, ...?". I think this should be done by the localization people, not the developer per se. Hence, the developer should just execute steps 1, 3 and 4 for a pull request.

**Comment by mlep** [comment]
I agree with @simonharrer .
In the meantime, if localization people could access automatic warnings, it would be nice.

**Comment by simonharrer** [comment]
What do you mean with automatic warnings? Which items have not been translated? Which keys are very similar?

**Comment by mlep** [comment]
I apologize for being unclear.
I was thinking that the item "Check whether keys can be reused, typos in keys, ...?" could benefit to localization people, even if they should not impeded the work of developers. Hence, could gradle or CircleCI generate warning messages? That way, we would know what need to be improved in the localization.
I hope it is clearer (but not sure...)

**Comment by simonharrer** [comment]
Hmm, can you provide examples for warning messages you would like to have, please? I think this would help. 

Developers would make sure that all keys within the Java code are present in all property files and that the english language file would be ok.

**Comment by mlep** [comment]
I was thinking about:
- list of localization files having keys waiting for translation
- space characters or tab characters present in the translated string

**Comment by matthiasgeiger** [comment]
> Who does "Check whether keys can be reused, typos in keys, ...?". I think this should be done by the localization people, not the developer per se.

Yes and no. ;-)

Why do I propose this? 
Because the time of the translators is valuable and translating hundreds of keys is not the most interesting work. So we developers should try to reduce the amount of work for the translators.

Stuff like "Found missing key: 'Thsi is a missng key'" should be clearly handled by the developer. 
"Found missing key: 'delete Entry'" - but there already exists a translated "delete entry" key - should also be possible for a developer to decide. 

But more complex reasoning often depends on language skills and is not per se decidable for developers. 

To sum up: Make life easier for translators and discipline developers regarding localization ;-)

**Comment by stefan-kolb** [comment]
I'd like to see this implemented in v3.1 :smile:
